graphql-language/src/parser.rs

```diff
@@ -0,0 +1,363 @@
+  lexer: Peekable<Lexer<'a>>,
+  source: Source<'a>,
+}
+
+impl<'a> Parser<'a> {
+  pub fn new(src: Source<'a>) -> Self {
+    Parser {
+      lexer: Lexer::new(src.body).peekable(),
+      source: src,
+    }
+  }
+
+  fn loc(&self, lo: usize, hi: usize) -> Location<'a> {
+    Location {
+      start: lo,
+      end: hi,
+      source: self.source,
+    }
+  }
+}
+
+impl<'a> Iterator for Parser<'a> {
+  type Item = Res<'a, Definition<'a>>;
+  fn next(&mut self) -> Option<Res<'a, Definition<'a>>> {
+    if self.lexer.peek().is_some() {
+      Some(parse_definition(self))
+    } else {
+      None
+    }
+  }
+}
+
+macro_rules! on_do {
+ ($parser: ident, $($p: pat => $b: block)+) => {
+   loop {
+     match next!($parser) {
+       $(
+         $p => $b
+       )+
+     }
+   }
+ }
+}
+
+macro_rules! peek {
+  ($parser: ident, $($p: pat)|*) => ({
+    let mut is_match = false;
+    if let &Ok(ref c) = $parser.lexer.peek().unwrap() {
+      match c {
+        $(
+          &$p => {
+            is_match = true;
+          }
+        )*
+        _ => {}
+      }
+    }
+    is_match
+  });
+}
+
+macro_rules! next {
+  ($parser: ident) => {$parser.lexer.next().unwrap().unwrap()}
+}
+
+fn parse_definition<'a>(parser: &mut Parser<'a>) -> Res<'a, Definition<'a>> {
+  let token = next!(parser);
+
+  match token {
+    Punctuator(LeftBrace, lo, _) => parse_short_operation(parser, lo),
+    TokenName(name, lo, hi) => {
+      match name {
+        "query" => parse_operation_def(parser, OperationType::Query, name, lo, hi),
+        "mutation" => parse_operation_def(parser, OperationType::Mutation, name, lo, hi),
+        "fragment" => parse_fragment_def(parser, name, lo, hi),
+        _ => Err("Unkown Op"),
+      }
+    },
+    _ => Err("Unexpected Token"),
+  }
+}
+
+fn parse_short_operation<'a>(parser: &mut Parser<'a>, lo: usize) -> Res<'a, Definition<'a>> {
+  let (selections, hi) = try!(parse_selection_set(parser, lo));
+
+  Ok(Definition::Operation(
+    OperationDefinition {
+      loc: parser.loc(lo, hi),
+      operation: OperationType::Query,
+      name: None,
+      variable_definitions: None,
+      directives: None,
+      selection_set: selections,
+    }
+  ))
+}
+
+fn parse_field<'a>(parser: &mut Parser<'a>, name: &'a str, lo: usize, mut hi: usize) -> DRes<'a, Selection<'a>> {
+  let (alias, name) = if peek!(parser, Punctuator(Colon, _, _)) {
+      let a = try!(parse_name(parser, name, lo, hi));
+      next!(parser);
+      (Some(a), match next!(parser) {
+        TokenName(n, l, h) => {
+          hi = h;
+          try!(parse_name(parser, n, l, h))
+        },
+        _ => {
+          return Err("Expected Name after colon");
+        },
+      })
+  } else {
+    (None, try!(parse_name(parser, name, lo, hi)))
+  };
+
+  let (args, hi) = try!(parse_arguments(parser, hi));
+  let (dirs, hi) = try!(parse_directives(parser, hi));
+  let (selections, hi) = try!(parse_selection_set(parser, hi));
+
+  Ok((Selection::Field(Field {
```

Comment:
Yeah, I think it would help me understand better to see `Ok(ast::Selection::Field(ast::Field { … }))`. My gut reaction was “wut?”